mlir/lib/Dialect/Rock/IR/RockDialect.cpp

```diff
@@ -588,12 +589,77 @@ static LogicalResult verifyGemmTypes(Operation *op, GemmFeatures features,
   return success();
 }
 
+static StringAttr getArchString(Operation *op) {
+  // Instantiate arch. It will get populated with the call to 'getAnyAttr' that
+  // takes place later
+  StringAttr arch;
+
+  // Lambda function to help with parsing 'arch' attributes from funcs
+  auto getAnyAttr = [&](ArrayRef<StringRef> attrNames, Operation *tempOp) {
+    for (StringRef attrName : attrNames) {
+      if (!arch) {
+        arch = tempOp->getAttrOfType<StringAttr>(attrName);
+      } else {
+        return;
+      }
+    }
+  };
+
+  Operation *func = op->getParentOfType<func::FuncOp>();
+  if (!func) {
+    func = op->getParentOfType<gpu::GPUFuncOp>();
+  }
+  getAnyAttr({"arch", "mhal.arch"}, func);
+
+  // If not available on the func, try looking at the module
+  if (!arch) {
+    auto mod = func->getParentOfType<ModuleOp>();
+    getAnyAttr({"arch", "mhal.arch"}, mod);
+  }
+
+  // If still not available, then we have malformed IR and need to error out
+  if (!arch)
+    llvm_unreachable("GEMM op needs to have an arch attribute");
+
+  return arch;
+}
+
+GemmFeatures getFeatureValue(Operation *op) {
```

Comment:
Perhaps create another header for this and use it in both loweringUtils.h and here ? instead of maintaning a copy